File: src/Builder.cc
#include <Builder.h>
#include <limits>
#include <iostream>
#include <cmath>

Builder::Builder(Data* data_){
    data = data_;
    typeSignal = 0;
    typeBackground = 1;
}

void Builder::build(Tree* tree){
    // Create the root node of the tree and attach all data to it
    tree->setRoot(new Node(0));
    for(unsigned int iSample=0; iSample<data->getNumSamples(); iSample++){
        tree->getRoot()->addSampleIndex(iSample);
    }

    // Start splitting from the root and proceed recursively until reaching
    // the desired depth of the tree
    split(tree, tree->getRoot());
}

void Builder::split(Tree* tree, Node* node){
    // Get histograms of features
    std::vector<std::vector<float> > hSig(data->getNumFeatures());
    std::vector<std::vector<float> > hBkg(data->getNumFeatures());
    for(unsigned int iFeature=0; iFeature<data->getNumFeatures(); iFeature++){
        hSig[iFeature].resize(data->getNumBins());
        hBkg[iFeature].resize(data->getNumBins());
    }
    std::vector<unsigned int>* sampleIndexes = node->getSampleIndexes();
    unsigned int iSample;
    for(unsigned int iSampleIndex=0; iSampleIndex<sampleIndexes->size(); iSampleIndex++){
        iSample = sampleIndexes->at(iSampleIndex);
        for(unsigned int iFeature=0; iFeature<data->getNumFeatures(); iFeature++){
            if(data->getType(iSample)==typeSignal) hSig[iFeature][data->getFeatureBin(iSample, iFeature)] += data->getWeight(iSample);
            else hBkg[iFeature][data->getFeatureBin(iSample, iFeature)] += data->getWeight(iSample);
        }
    }

    // Turn the histograms into cumulative histograms
    for(unsigned int iFeature=0; iFeature<data->getNumFeatures(); iFeature++){
        for(unsigned int iBin=1; iBin<data->getNumBins(); iBin++){
            hSig[iFeature][iBin] += hSig[iFeature][iBin-1];
            hBkg[iFeature][iBin] += hBkg[iFeature][iBin-1];
        }
    }

    // Calculate the separation gain for each bin and feature by finding the minimal loss.
    // This gives the best cut on the features for the given binning.
    std::cout << "Layer: " << node->getLayer() << std::endl;
    float minLoss = std::numeric_limits<float>::max();
    unsigned int bestBin, bestFeature;
    float purityLeft, purityRight, loss;
    for(unsigned int iFeature=0; iFeature<data->getNumFeatures(); iFeature++){
        for(unsigned int iBin=0; iBin<data->getNumBins(); iBin++){
            purityLeft = hSig[iFeature][iBin]/(hSig[iFeature][iBin]+hBkg[iFeature][iBin]);
            purityRight = (hSig[iFeature].back()-hSig[iFeature][iBin])/(hSig[iFeature].back()-hSig[iFeature][iBin]+hBkg[iFeature].back()-hBkg[iFeature][iBin]);
            loss = purityLeft*(1.0-purityLeft) + purityRight*(1-purityRight);
            if(std::isnan(loss)) loss = std::numeric_limits<float>::max();
            if(loss<minLoss){
                minLoss = loss;
                bestBin = iBin;
                bestFeature = iFeature;
            }
        }
    }

    // Make this node a leaf node if the cutting failed.
    // As well, make this node a leaf node if the depth of the tree
    // is reached.
    // Determine to node type by choosing the greater number of signal or
    // background.
    if(node->getLayer()==tree->getDepth()){
        node->setLeaf(true);
        unsigned int sumSig = 0;
        unsigned int sumBkg = 0;
        for(unsigned int iSampleIndex=0; iSampleIndex<sampleIndexes->size(); iSampleIndex++){
            iSample = sampleIndexes->at(iSampleIndex);
            if(data->getType(iSample)==typeSignal) sumSig++;
            else sumBkg++;
        }
        if(sumSig>sumBkg) node->setType(typeSignal);
        else node->setType(typeBackground);
        return;
    }

    std::cout << "Best feature/bin: " << bestFeature << " " << bestBin << std::endl;

    // Otherwise, set the cut values and split into two new nodes in the next
    // layer and append the appropriate data to them.
    // A node on the right contains all values which are greater than the cut.
    node->setCutFeature(bestFeature);
    node->setCutBin(bestBin);
    node->setCutValue(data->getValueFromBin(bestFeature, bestBin));

    Node* right = new Node(node->getLayer()+1);
    Node* left = new Node(node->getLayer()+1);
    node->setRight(right);
    node->setLeft(left);
    for(unsigned int iSampleIndex=0; iSampleIndex<sampleIndexes->size(); iSampleIndex++){
        iSample = sampleIndexes->at(iSampleIndex);
        if(data->getFeatureBin(iSample, bestFeature)>bestBin) right->addSampleIndex(iSample);
        else left->addSampleIndex(iSample);
    }

    std::vector<unsigned int>* x = right->getSampleIndexes();
    std::vector<unsigned int>* y = left->getSampleIndexes();
    std::cout << "Right/left: " << x->size() << " " << y->size() << std::endl;

    split(tree, right);
    split(tree, left);
}

File: python/Builder.py
from Tree import Tree
from Node import Node
import numpy as np

def getMin(eventSample):
    mins = []
    for iFeature in range(len(eventSample[0][0])):
        mins.append(1000)
        for event in eventSample:
            if event[0][iFeature]<mins[iFeature]:
                mins[iFeature] = event[0][iFeature]
    return mins

def getMax(eventSample):
    maxs = []
    for iFeature in range(len(eventSample[0][0])):
        maxs.append(-1000)
        for event in eventSample:
            if event[0][iFeature]>maxs[iFeature]:
                maxs[iFeature] = event[0][iFeature]
    return maxs

class Builder:
    def __init__(self):
        self.eventSample = None

    def build(self, eventSample, treeDepth, numCuts):
        self.treeDepth = treeDepth
        self.numCuts = numCuts
        # Get tree
        tree = Tree(treeDepth)
        # Set root node of tree
        tree.root = Node(0, eventSample)
        # Split the node recursively
        self.split(tree.root)
        return tree

    def split(self, node):
        # Make this node a leaf if depth of tree is reached
        if node.layer==self.treeDepth:
            node.setLeafClass()
            return
        # Get the cut values of the features
        cuts = []
        mins = getMin(node.eventSample)
        maxs = getMax(node.eventSample)
        for i in range(len(mins)):
            cuts.append(np.linspace(mins[i], maxs[i], self.numCuts))
        # Find the best cut
        minLoss = 1000
        bestCut = None
        bestFeature = None
        for iFeature in range(len(cuts)):
            for iCut in range(len(cuts[0])):
                numSigLeft = 0
                numBkgLeft = 0
                numSigRight = 0
                numBkgRight = 0
                for iSample in range(len(node.eventSample)):
                    features = node.eventSample[iSample][0]
                    weight = node.eventSample[iSample][1]
                    class_ = node.eventSample[iSample][2]
                    if features[iFeature]>cuts[iFeature][iCut]:
                        if class_ == 1:
                            numSigRight += weight
                        else:
                            numBkgRight += weight
                    else:
                        if class_ == 1:
                            numSigLeft += weight
                        else:
                            numBkgLeft += weight
                purityLeft = None
                if numSigLeft+numBkgLeft>0:
                    purityLeft = numSigLeft/(numSigLeft+numBkgLeft)
                purityRight = None
                if numSigRight+numBkgRight>0:
                    purityRight = numSigRight/(numSigRight+numBkgRight)
                loss = None
                if purityLeft!=None and purityRight!=None:
                    loss = purityLeft*(1.0-purityLeft) + purityRight*(1.0-purityRight)
                    #print("%i\t%.2f\t%.2f" % (iFeature, cuts[iFeature][iCut], loss))
                if loss!=None:
                    if loss<minLoss:
                        minLoss = loss
                        bestCut = cuts[iFeature][iCut]
                        bestFeature = iFeature
        # Set cut value for this node
        node.cutValue = bestCut
        node.cutFeature = bestFeature
        #print("Feature:", bestFeature)
        #print("Cut:", bestCut)
        # Split the node
        eventSampleRight = []
        eventSampleLeft = []
        for sample in node.eventSample:
            if sample[0][bestFeature]>bestCut:
                eventSampleRight.append(sample)
            else:
                eventSampleLeft.append(sample)
        node.right = Node(node.layer+1, eventSampleRight)
        node.left = Node(node.layer+1, eventSampleLeft)
        # Split the left and right nodes again
        self.split(node.right)
        self.split(node.left)
